dora/core/client/hdfs/src/main/java/alluxio/hadoop/AbstractFileSystem.java

```diff
@@ -325,7 +325,7 @@ public BlockLocation[] getFileBlockLocations(FileStatus file, long start, long l
               info.getBlockInfo().getLength()));
         }
       });
-      BlockLocation[] ret = blockLocations.toArray(new BlockLocation[blockLocations.size()]);
+      BlockLocation[] ret = blockLocations.toArray(new BlockLocation[0]);
```

Comment:
The docs says:

```
Suppose x is a list known to contain only strings. The following code can be used to dump the list into a newly allocated array of String:


     String[] y = x.toArray(new String[0]);
 
Note that toArray(new Object[0]) is identical in function to toArray().


```

So, maybe we can use the .toArray() instead?

Reply:
`toArray` needs a typed argument to do type inference. If no such argument is provided, it returns an untyped array which is of type `Object[]`.

Reply:
I see. 
But, `blockLocations` has already its type defined when it is declared. Maybe this Java.List can not get such type information.

Reply:
Above is correct. At least in Java 8, `toArray(new XXX[0])` is faster than `toArray(new XXX(size))`. I don't know if that's still the case in java 11+.

Also, in terms of this PR, I only care if I'm doing the cherry-pick correctly. I don't want to make the review longer because we want to re-consider the code change made in any of the commits.

Your comment is still highly valued though, feel free to leave all sorts of comments like that! It's just I may not address them. If you feel some changes can be made, feel free to just create separate PRs into `main` to make improvements :)

Reply:
You are absolutely right; this is a limitation of Java™ generics.